staticaddr/withdraw/manager.go

```diff
@@ -0,0 +1,726 @@
+		return err
+	}
+
+	m.Lock()
+	m.activeWithdrawals[withdrawalAddress.String()] = deposits
+	m.Unlock()
+
+	return nil
+}
+
+func (m *Manager) publishWithdrawal(ctx context.Context,
+	deposits []*deposit.Deposit, withdrawalAddress btcutil.Address) error {
+
+	// Create a musig2 session for each deposit.
+	withdrawalSessions, clientNonces, err := m.createMusig2Sessions(
+		ctx, deposits,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Get the fee rate for the withdrawal sweep.
+	withdrawalSweepFeeRate, err := m.cfg.WalletKit.EstimateFeeRate(
+		m.runCtx, defaultConfTarget,
+	)
+	if err != nil {
+		return err
+	}
+
+	outpoints := toOutpoints(deposits)
+	resp, err := m.cfg.StaticAddressServerClient.ServerWithdrawDeposits(
+		m.runCtx, &staticaddressrpc.ServerWithdrawRequest{
+			Outpoints:       toServerOutpoints(outpoints),
+			ClientNonces:    clientNonces,
+			ClientSweepAddr: withdrawalAddress.String(),
+			MusigTxFeeRate:  uint64(withdrawalSweepFeeRate),
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	addressParams, err := m.cfg.AddressManager.GetStaticAddressParameters(
+		ctx,
+	)
+	if err != nil {
+		return fmt.Errorf("couldn't get confirmation height for "+
+			"deposit, %v", err)
+	}
+
+	prevOuts := m.toPrevOuts(deposits, addressParams.PkScript)
+	totalValue := withdrawalValue(prevOuts)
+	withdrawalTx, err := m.createWithdrawalTx(
+		outpoints, totalValue, withdrawalAddress,
+		withdrawalSweepFeeRate,
+	)
+	if err != nil {
+		return err
+	}
+
+	coopServerNonces, err := toNonces(resp.ServerNonces)
+	if err != nil {
+		return err
+	}
+
+	// Next we'll get our sweep tx signatures.
+	prevOutFetcher := txscript.NewMultiPrevOutFetcher(prevOuts)
+	_, err = m.signMusig2Tx(
+		m.runCtx, prevOutFetcher, outpoints, m.cfg.Signer, withdrawalTx,
+		withdrawalSessions, coopServerNonces,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Now we'll finalize the sweepless sweep transaction.
+	finalizedWithdrawalTx, err := m.finalizeMusig2Transaction(
+		m.runCtx, outpoints, m.cfg.Signer, withdrawalSessions,
+		withdrawalTx, resp.Musig2SweepSigs,
+	)
+	if err != nil {
+		return err
+	}
+
+	txLabel := fmt.Sprintf("deposit-withdrawal-%v",
+		finalizedWithdrawalTx.TxHash())
+
+	// Publish the withdrawal sweep transaction.
+	err = m.cfg.WalletKit.PublishTransaction(
+		m.runCtx, finalizedWithdrawalTx, txLabel,
+	)
+
+	if err != nil {
+		if !strings.Contains(err.Error(), "output already spent") {
+			log.Errorf("%v: %v", txLabel, err)
+			return err
+		}
+	} else {
+		log.Debugf("published deposit withdrawal with txid: %v",
+			finalizedWithdrawalTx.TxHash())
+	}
+
+	return err
+}
+
+func (m *Manager) handleWithdrawal(deposits []*deposit.Deposit,
+	withdrawalAddress btcutil.Address) error {
+
+	withdrawalPkScript, err := txscript.PayToAddrScript(withdrawalAddress)
+	if err != nil {
+		return err
+	}
+
+	confChan, errChan, err := m.cfg.ChainNotifier.RegisterConfirmationsNtfn(
+		m.runCtx, nil, withdrawalPkScript, MinConfs,
+		int32(m.initiationHeight),
```

Comment:
Access to initiationHeight without a lock.

I suggest to use [atomic.Uint32](https://pkg.go.dev/sync/atomic@go1.19.10#Uint32) for initiationHeight and currentHeight.

Reply:
On the other hand I think it is nice that all state is locked together. I think it's probably better to keep so and not introduce atomics for these specific use cases.